### ISSUE TITLE
Circular references while using Pick

### ISSUE THREAD

**Issue by nscarcella** [microsoft/TypeScript#32986]
Hello everybody,

I found a weird behavior while trying to make a type that would make all (...| undefined) optional in a cyclic model. It seems like using the Pick type triggers a "Type alias circularly references itself" error on models that otherwise resolve fine.

The weirdest part is that the problem seems to be fixable by replacing Pick with what, I believe, is its current definition.

Here is a simplified example of my current model:

```ts
type NonOptionalKeys<N> = {[K in keyof N]: undefined extends N[K] ? never : K}[keyof N]

// This definition of Optional fails
type Optional<N> = Partial<N> & Pick<N, NonOptionalKeys<N>>

// While this one seems to behave fine
// type Optional<N> = Partial<N> & { [P in NonOptionalKeys<N>]: N[P] }

type Expression<T> = ExpressionA<T> | ExpressionB<T>

type ExpressionA<T> = Optional<{
    x: T
}>

type ExpressionB<T> = Optional<{
    y: Expression<T>
    z: T
}>
```

I know I'm stepping on thin ice with the whole circular references thing and maybe there is an explanation but I'm leaving the report just in case it helps.

**Search Terms:**
Circular reference, Pick, Cyclic

**Expected behavior:**
The Pick doesn't affect the circular reference resolution

**Actual behavior:**
Pick seems to cause a false positive

[**Playground Link**](https://www.typescriptlang.org/play/#code/C4TwDgpgBAcg9gOwPJmAS0QQwDYGkIgDOAPDAHxQC8UA3gNq5RoJQDWBcAZrALoBcUAK4IAJhE7MIIqBAAewCKMKwGPKAH4oCCADcIAJygDcAXzrsQXXgFgAUHYD0DqABUAFmmViJCNOkRQVij+CDhQnJho2IR2oJBQwRih2KQU1AAKmProOKlQAGRQ6WgAxqykADSwiIlYeAQk5GR2js4A6h7Y0MAeyojQhBAQALbKwHBQAEYQbph64ZKtUHHQtcl5GVk5KeQFtFB06Uws8MioSTj4RKn8KulqJi22K1AAorJg+hCEhEnELmk3h8vj8kgBBf4UAA+QM+31+iAAQpCni93nDQYgIQCqAlznViDQ7FASVBZAIXHYTM17M9wNB0SCEQhkTjqGtckTbKSoCABIz4X8AcTSQAvClUshAA)

**Comment by nscarcella** [comment]
Update: Testing this on 3.5.3 I get a different error for both versions:

Type parameter 'P' has a circular constraint

**Comment by weswigham** [comment]
Good news OP! Both of your examples now behave the same - they issue a circularity error. The optionality of `ExpressionB`'s keys depends on evaluating the `Expression` type, which in turn contains `ExpressionB`. I'm speculating, but we used to not notice the circular structure in the case of the object literal because we previously didn't recognize that object literal as "homomorphic" over the input type (so we didn't evaluate it memberwise to calculate the optionality), whereas nowadays we do. In any case, I don't think we have anything to do here anymore.
